### DISCUSSION TITLE
Rust-analyzer will not initialize when using rustaceanvim

### DISCUSSION THREAD

**Discussion by TrentDaniel** [mrcjkb/rustaceanvim#469] [Q&A]
# Description
I'm trying to setup rustaceanvim so that I can use the debugger with my Rust projects. I was able to register the rust-analyzer as a listed server and attach it to Rust file buffers when using rust-tools, but I got a ton of errors as soon as I tried using the debugger, so I decided to move onto rustaceanvim since it is still maintained (thanks btw). While using rustaceanvim, I have not been able to get the rust-analyzer or any other rust related string to appear in the list of registered servers, or to display as an attached client, as shown below:
```
Language client log: ~/.local/state/nvim/lsp.log
Detected filetype: rust

0 client(s) attached to this buffer: <-- using rust-tools, this said that there was 1 client attached to the buffer

Configured servers list: lua_ls  <-- using rust-tools, this had an additional entry called 'rust-analyzer'
```
The above output is from running `:LspInfo` while having a `.rs` file open.

My only goals atm are:
- [ ] have rustaceanvim activate rust-analyzer as an LSP for .rs files
- [ ] be able to use the debugger etc. through rustaceanvim for .rs files
- [ ] Gain a better understanding of how rustaceanvim is configured etc. so that I can solve these problems more easily on my own

# Environment
I am using Ubuntu 24.04 inside a docker container, and neovim 0.9.5. I have not been able to install neovim 0.10.0 on my ubuntu distro yet. Neovim 0.10.0 has been listed as a requirement, so I'm not sure if maybe that's the problem??
I have no idea what version of NvChad or rustaceanvim I'm using exactly, because I don't know how to check for that either from the terminal CLI or through editor commands.

# rust-analyzer installation
I have verified that I have rust-analyzer installed through cargo, and that it is on my path:
```
> which rust-analyzer
~/.cargo/bin/rust-analyzer
```
I have verified that rust-analyzer has not been installed by Mason:
```
> ls ~/.local/share/nvim/mason/packages
lua-language-server
```
I have verified that rustaceanvim should have all that it needs through a healthcheck:
```
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies
- OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.

Checking external dependencies
- OK rust-analyzer: found rust-analyzer 1.80.0 (0514789 2024-07-21)
- OK Cargo: found cargo 1.80.0 (376290515 2024-07-16)
- OK rustc: found rustc 1.80.0 (051478957 2024-07-21)

Checking config
- OK No errors found in config.

Checking for conflicting plugins
- OK No conflicting plugins detected.

Checking for tree-sitter parser
- OK tree-sitter parser for Rust detected.
```
The above healthcheck also verifies that the rust-analyzer being considered by rustaceanvim is one installed through rustup and not by Mason.

# NvChad Configurations
I'm using NvChad to support my editor, with the following file structure under `~/.config/nvim`:
```
.git
.github
lua
    - core
        - bootstrap.lua
        - default_config.lua
        - init.lua
        - mappings.lua
        - utils.lua
    - custom
        - configs
            - lspconfig.lua
            - rust-tools.lua
            - rustaceanvim.lua
        - chadrc.lua
        - mappings.lua
        - plugins.lua
    - plugins
        - configs
            - cmp.lua
            - lazy_nvim.lua
            - lspconfig.lua
            - mason.lua
            - nvimtree.lua
            - others.lua
            - telescope.lua
            - treesitter.lua
        - init.lua

.gitignore
.ignore
.stylua.toml
init.lua
lazy-lock.json
LICENSE
```

I haven't manually changed any of the files in the `lua/core` or `lua/plugins` directories, so at least for now we can assume that those have retained the default implementations.

While I'm pretty sure this file is no longer being meaningfully used since the previous server registration for rust-analyzer contained in this file has been commented out, the contents of `lua/custom/configs/lspconfig.lua` is:
```
local on_attach = require("plugins.configs.lspconfig").on_attach
local capabilities = require("plugins.configs.lspconfig").capabilities
local lspconfig = require("lspconfig")
local util = require"lspconfig/util"
```

While I'm pretty sure this file is no longer being imported/used anywhere since I removed the rust-tools plugin from the `plugins.lua` file, the contents of `lua/custom/configs/lspconfig.lua` is:
```
local on_attach = require("plugins.configs.lspconfig").on_attach
local capabilities = require("plugins.configs.lspconfig").capabilities

local options = {
    server = {
        on_attach = on_attach,
        capabilities = capabilities,
    }
}

return options
```

The contents of the `lua/custom/configs/rustaceanvim.lua` is:
```
local on_attach = require("plugins.configs.lspconfigs").on_attach
local capabilities = require("plugins.configs.lspconfig").capabilities
local util = require"lspconfig/util"

vim.g.rustaceanvim = {
    server = {
        on_attach = on_attach,
        capabilities = capabilities,
        filetypes = {"rust"},
        root_dir = util.root_pattern("Cargo.toml"),
        settings = {
            ['rust-analyzer'] = {
                cargo = {
                    allFeatures = true,
                },
            },
        },
    }
}
```

The contents of `lua/custom/chadrc.lua is:
```
--@type ChadrcConfig
local M = {}
M.ui = { theme = 'catppuccin' }
M.plugins = 'custom.plugins'
M.mappings = require "custom.mappings"
return M
```

The contents of `lua/cusom/mappings.lua` is:
```
local M = {}

M.dap = {
    plugin = true,
    n = {
        ["<leader>db"] = { "<cmd> DapToggleBreakpoint <CR>" },
        ["<leader>dus"] = {
            function ()
                local widgets = require('dap.ui.widgets');
                local sidebar = widgets.sidebar(widgets.scopes);
                sidebar.open();
            end
            "Open debugging sidebar"
        }
    }
}

return M
```

The contents of `lua/custom/plugins.lua` is:
```
local cmp = require "cmp"

local plugins = {
    {
        "williamboman/mason.nvim",
    },
    {
        "neovim/.nvim-lspconfig",
        config = function()
            require "plugins.configs.lspconfig"
            require "custom.configs.lspconfig"
        end,
    },
    {
        "rust-lang/rust.vim",
        ft = "rust",
        init = function ()
            vim.g.rustfmt_autosave = 1
        end
    },
    {
        "mrcjkb/rustaceanvim",
        version = "^5",
        ft = { "rust" },
        config = function()
            require "custom.configs.rustaceanvim"
        end,
    },
    {
        "mfussenegger/nvim-dap",
        init = function()
            require("core.utils").load_mappings("dap")
        end,
    },
}

return plugins
```

# Note
I understand that there is probably too much information included in this question, but I've scoured google etc. for information on this problem and haven't found any so I have no idea where the problem might be. This makes it difficult for me to narrow the amount of information provided above. I would really appreciate your help in fixing this problem!

**Comment by mrcjkb** [comment]
Hey :wave: 

I don't have the capacity to provide support for all the distributions out there.
Please search the existing discussions for NVChad, and also search the NVChad issues + discussions.

Marking as duplicate of: https://github.com/mrcjkb/rustaceanvim/discussions/351#discussioncomment-9097402
